xapian-core/cluster/kmeans.cc

```diff
@@ -24,120 +24,126 @@
 #include "xapian/cluster.h"
 #include <debuglog.h>
 #include <xapian/error.h>
+#include <omassert.h>
 
-#include <time.h>
-#include <cstdlib>
 #include <climits>
+#include <random>
 #include <set>
 #include <unordered_map>
 #include <vector>
 
+/** Threshold value for checking convergence in KMeans
+ */
 #define EPSILON 0.0000000001
 
 using namespace Xapian;
 using namespace std;
 
+KMeans::KMeans(unsigned int k_) {
+    LOGCALL_CTOR(API, "KMeans()", k_);
+    if (k == 0)
+	throw InvalidArgumentError("Number of required clusters should be greater than zero");
+    k = k_;
+    max_iters = 100;
+}
+
+KMeans::KMeans(unsigned int k_, unsigned int max_iters_) {
+    LOGCALL_CTOR(API, "KMeans()", k_ | max_iters_);
+    if (k == 0)
+	throw InvalidArgumentError("Number of required clusters should be greater than zero");
+    k = k_;
+    max_iters = max_iters_;
+}
+
 string
 KMeans::get_description() const {
     LOGCALL(API, string, "KMeans::get_description()", NO_ARGS);
     return "KMeans clusterer";
 }
 
 void
-KMeans::initialize_centroids(ClusterSet &cset) {
-    LOGCALL_VOID(API, "KMeans::initialize_centroids()", cset);
-    set<docid> centroid_docs;
+KMeans::initialize_clusters(ClusterSet &cset) {
+    LOGCALL_VOID(API, "KMeans::initialize_clusters()", cset);
     int size = docs.size();
-    srand(time(NULL));
+    mt19937 generator;
+    generator.seed(random_device()());
+    uniform_int_distribution<mt19937::result_type> random(0, size);
     for (unsigned int i = 0; i < k;) {
-	int x = rand() % size;
-	if (centroid_docs.find(x) == centroid_docs.end()) {
-	    Point a = docs[x];
-	    Centroid centroid;
-	    centroid.set_to_point(a);
-	    centroids.push_back(centroid);
-	    centroid_docs.insert(x);
-	    Cluster cluster1(centroid);
-	    cset.add_cluster(cluster1);
-	    ++i;
-	}
+	int x = random(generator);
+	Centroid centroid(docs[x]);
+	Cluster cluster(centroid);
+	cset.add_cluster(cluster);
+	++i;
```

Comment:
Since the increment is now unconditional, you can put it in the `for` statement. I think this warrants a comment explaining why you don't have to check for duplicated centroids (on the basis of the long period of mt19937).